### DISCUSSION TITLE
How to get Tailwind CSS IntelliSense working

### DISCUSSION THREAD

**Discussion by woller** [yippee-fun/phlex#536] [Help] [answered]
I'm playing around with Phlex and I can really see some potential here. It's a very cool project. However the fact, that I cannot get the Tailwind CSS IntelliSense VSCode extension working, is holding me back from wanting to use it. This may very well be just me, who cannot figure out how to configure it correctly 🙂 
Has anyone managed to get IntelliSense working properly in phlex views and components?

**Comment by joeldrapper** [comment]
I expect Tailwind auto-complete extensions use an HTML parser. They would need to understand Ruby to work with Phlex. I think it should be possible to develop better editor tooling for Phlex (including Tailwind support) as an extension for the [Ruby LSP](https://github.com/Shopify/ruby-lsp) project in the future. I don't think the extension API is ready yet.

**Reply by woller** [reply]
That's too bad. I was hoping it was just me not understanding something 🙂 But thank you for the answer and for a really cool gem

**Reply by joeldrapper** [reply]
You’re welcome. I do hope to be able to have a Ruby LSP plugin with Tailwind support soon.

**Comment by hangsu** [comment]
I think the Tailwind auto-complete extension actually uses regex.

I haven't tested this with VSCode, but you should be able to use the `tailwindCSS.experimental.classRegex` [configuration](https://github.com/tailwindlabs/tailwindcss-intellisense/blob/89dd1044190e1b713c8108d50d0a2fe99a0f8760/packages/tailwindcss-language-service/src/completionProvider.ts#L481) to make it work.

Might need to mess around with `tailwindCSS.includeLanguages` too.

**Reply by joeldrapper** [reply]
That looks promising. I’m still very interested in building a Ruby LSP plugin for Phlex and Phlex/Rails but this might be a good compromise for now.

**Comment by zealot128** [comment, marked as answer]
Made it work - using Lsp Config in Neovim and played around a little:

https://gist.github.com/zealot128/cf8b0ad74df5a9ab3e734b9e6acc41e8


```lua
 filetypes = { "aspnetcorerazor", "astro", "astro-markdown", "blade", "clojure", "django-html", "htmldjango", "edge", "eelixir", "elixir", "ejs", "erb", "eruby", "gohtml", "haml", "handlebars", "hbs", "html", "html-eex", "heex", "jade", "leaf", "liquid", "markdown", "mdx", "mustache", "njk", "nunjucks", "php", "razor", "slim", "twig", "css", "less", "postcss", "sass", "scss",   "stylus", "sugarss", "javascript", "javascriptreact", "reason", "rescript", "typescript", "typescriptreact", "vue", "svelte", "ruby" },  
  init_options = {        
    userLanguages = {
      ruby = "php",
    },
  },
  settings = {
    tailwindCSS = {
      experimental = {
        classRegex = {
    -- add class: "..." and class: '' 
          [[class= "([^"]*)]],
          [[class: "([^"]*)]], 
          [[class= '([^']*)]],
          [[class: '([^']*)]],
         -- 
         --  '~H""".*class="([^"]*)".*"""',
         --  '~F""".*class="([^"]*)".*"""',  
        },
      } 
    }
  }
```

- map ruby to php (or any other language that is listed in the html languages)
- add ruby to the filetypes so the LSP activates (Neovim)
- clss Regexp, not sure what the last 2 lines are doing, that's borrowed from the current issue https://github.com/tailwindlabs/tailwindcss-intellisense/issues/737, so just comment

KISS regexp parsing can do most of the job I guess.. :)

![grafik](https://user-images.githubusercontent.com/147175/230392281-30acb4cc-4004-4147-92f0-a358e9442564.png)

**Reply by joeldrapper** [reply]
This looks great!

**Reply by zealot128** [reply]
Updates answer for tailwind newer lsps (0.14):

```lua
      lspconfig.tailwindcss.setup {
        on_attach = on_attach,
        flags = lsp_flags,
        filetypes = { "aspnetcorerazor", "astro", "astro-markdown", "blade", "clojure", "django-html", "htmldjango", "edge",
          "eelixir", "elixir", "ejs", "erb", "eruby", "gohtml", "haml", "handlebars", "hbs", "html", "html-eex", "heex",
          "jade", "leaf", "liquid", "markdown", "mdx", "mustache", "njk", "nunjucks", "php", "razor", "slim", "twig", "css",
          "less", "postcss", "sass", "scss", "stylus", "sugarss", "javascript", "javascriptreact", "reason", "rescript",
          "typescript", "typescriptreact", "vue", "svelte", "ruby" },
        settings = {
          tailwindCSS = {
            includeLanguages = {
              ruby = "php"
            },
            experimental = {
              classRegex = {
                [[class= "([^"]*)]],
                [[class: "([^"]*)]],
                [[class= '([^']*)]],
                [[class: '([^']*)]],
                '~H""".*class="([^"]*)".*"""',
                '~F""".*class="([^"]*)".*"""',
              },
            }
          }
        }
}
```

classRegex is still required, but there is a new includeLanguages attribute.

**Comment by bbugh** [comment]
Here is the configuration for VSCode using the same tailwind-intellisense's `experimental.classRegex` setting that zeal128 mentioned above, with improvements/corrections to the regex.

You need to first turn it on for Ruby using `includeLanguages` (which isn't needed for `erb`, as tailwind is already enabled for that). Then specify the `classRegex`. This will make `class: ""` or `class: ''` syntax work anywhere in any language:

```json
{
  "tailwindCSS.includeLanguages": {
    "ruby": "html"
  },
  "tailwindCSS.experimental.classRegex": [
	"\\bclass:\\s*[\"']([^\"']*)[\"']",
  ]
}
```

If you want to only enable it only for the Ruby language for safety, you can specify it in the [ruby] settings block, (and [erb] if you want that too).

```json
{
  "tailwindCSS.includeLanguages": {
    "ruby": "html"
  },
  "[ruby]":{
    "tailwindCSS.experimental.classRegex": [
      "\\bclass:\\s*[\"']([^\"']*)[\"']",
    ]
  },
  "[erb]":{
    "tailwindCSS.experimental.classRegex": [
      "\\bclass:\\s*[\"']([^\"']*)[\"']",
    ]
  }
}
```

**Reply by joeldrapper** [reply]
🤩 I wonder if this works with Zed. 🤔

**Reply by pjmelling** [reply]
Nice, this worked for me. Thank you!
